modules/katamari/src/main/ts/ephox/katamari/api/Maybes.ts

```diff
@@ -0,0 +1,330 @@
+// output will also contain a value (that value being the output of
+// `mapper(self.value)`), and if self **does not** contain a value then
+// neither will the output.
+export const map = <T, U>(mapper: (value: T) => U) => (self: Maybe<T>): Maybe<U> =>
+  self.tag === Tag.Just ? just(mapper(self.value)) : nothing();
+
+// --- Applicative ---
+
+// Perform a transform on two Maybe objects, **if** they both have values. If
+// either of the objects does not contain a value, this will return an empty
+// Maybe (nothing), but if all of the input objects do have a value then `fn`
+// will be called (with all of those values as its arguments) and the output
+// will be wrapped in another Maybe.
+export const lift2 = <A, B, Out>(a: Maybe<A>, b: Maybe<B>, fn: (a: A, b: B) => Out): Maybe<Out> => {
+  if (a.tag === Tag.Just && b.tag === Tag.Just) {
+    return just(fn(a.value, b.value));
+  } else {
+    return nothing<Out>();
+  }
+};
+
+// Same as lift2, but with 3 maybe objects.
+export const lift3 = <A, B, C, Out>(a: Maybe<A>, b: Maybe<B>, c: Maybe<C>, fn: (a: A, b: B, c: C) => Out): Maybe<Out> => {
+  if (a.tag === Tag.Just && b.tag === Tag.Just && c.tag === Tag.Just) {
+    return just(fn(a.value, b.value, c.value));
+  } else {
+    return nothing<Out>();
+  }
+};
+
+// Same as lift2, but with 4 maybe objects.
+export const lift4 = <A, B, C, D, Out>(
+  a: Maybe<A>,
+  b: Maybe<B>,
+  c: Maybe<C>,
+  d: Maybe<D>,
+  fn: (a: A, b: B, c: C, d: D) => Out
+): Maybe<Out> => {
+  if (a.tag === Tag.Just && b.tag === Tag.Just && c.tag === Tag.Just && d.tag === Tag.Just) {
+    return just(fn(a.value, b.value, c.value, d.value));
+  } else {
+    return nothing<Out>();
+  }
+};
+
+// Same as lift2, but with 5 maybe objects.
+export const lift5 = <A, B, C, D, E, Out>(
+  a: Maybe<A>,
+  b: Maybe<B>,
+  c: Maybe<C>,
+  d: Maybe<D>,
+  e: Maybe<E>,
+  fn: (a: A, b: B, c: C, d: D, e: E) => Out
+): Maybe<Out> => {
+  if (a.tag === Tag.Just && b.tag === Tag.Just && c.tag === Tag.Just && d.tag === Tag.Just && e.tag === Tag.Just) {
+    return just(fn(a.value, b.value, c.value, d.value, e.value));
+  } else {
+    return nothing<Out>();
+  }
+};
+
+// --- Monad ---
+
+// Un-nests a Maybe type. Conversion is as follows:
+// Nothing ==> Nothing
+// Just(Nothing) ==> Nothing
+// Just(Just(T)) ==> Just(T)
+//
+// You shouldn't often end up with nested Maybe types to need this function, but
+// if you do it's here.
+export const flatten = <T>(self: Maybe<Maybe<T>>): Maybe<T> =>
+  self.tag === Tag.Just ? self.value : nothing();
+
+// Perform a transform on a Maybe object, **if** there is a value. Unlike the
+// map function earlier in the piece, here the transform itself also returns a
+// Maybe. Think of this like a combination of `map` and `flatten` at the same
+// time.
+export const bind = <T, U>(binder: (value: T) => Maybe<U>) => (self: Maybe<T>): Maybe<U> =>
+  self.tag === Tag.Just ? binder(self.value) : nothing();
+
+// --- Traversable ---
```

Comment:
Traversible means something different to this.

Reply:
Actually, these functions are from Foldable, which is a superclass of Traversable, so it's fine.

Reply:
`Maybe` is Travesable (since it's a Functor and a Foldable). There might be some other useful functions we could implement there.